### PR TITLE
QVAC-23758 fix: drain queued streaming output before JobEnded in parakeet endStreaming

### PR DESCRIPTION
**Pull request by ogad-tether**
## Summary

Fixes the Parakeet duplex-streaming tail-cut reported in QVAC-23758 ([Asana](https://app.asana.com/1/45238840754660/project/1214153063536860/task/1217548655428314)): when the audio input stream ended while the engine still had a buffered backlog, the end of the transcript was silently dropped.

> **Note:** originally implemented against `packages/transcription-parakeet`; rebased and ported to `packages/asr-ggml` after the whisper/parakeet unification (#3553) landed on main. The bug carried over verbatim into the unified package's parakeet driver.

### Root cause

The native `endStreaming` joins the worker thread only **after** it drains all buffered audio and finalizes — every tail segment is queued into the addon's output queue correctly. But delivery of those queued events to JS is asynchronous (`uv_async`), and the parakeet driver (`src/engines/parakeet/parakeet.ts`) cleared `_activeJobId` and emitted a synthetic `JobEnded` synchronously right after the native call returned. When the drained segments arrived at `_addonOutputCallback`, `jobId === null` → all discarded. Apps feeding faster than realtime (e.g. the SDK's `transcribeStream` fed from a file, then `end()`) lost most of the unprocessed tail.

### Fix (mirrors the whisper engine's `StreamingProcessor`)

- **`addon/src/model-interface/ParakeetStreamingProcessor.cpp`** — after `finalize()` on the graceful end path, queue a terminal `RuntimeStats` (real `audioDurationMs` / `totalSamples` from the processor's own counters — `ParakeetModel::runtimeStats()` is bypassed by the duplex path) through the **same FIFO output queue** as the drained segments, so it can never outrun them. This is exactly where whisper's `StreamingProcessor` calls `queueJobEnded()`.
- **`src/engines/parakeet/parakeet.ts`** — `endStreaming()` no longer clears the job or synthesises a `JobEnded`; it waits for the queue-delivered terminal event to flow through the normal output-callback path, so every drained segment is delivered in order before the response resolves. The synthetic `JobEnded` remains only as a fallback when no native session existed (`cleaned: false`). Waiter resolution is also wired into the cancel/destroy paths so nothing can hang. Generated `engines/parakeet/parakeet.{js,d.ts}` rebuilt via `build:ts` (`check:generated` passes).
- **`test/mocks/ParakeetMockedBinding.js`** — mimics the new contract (graceful end keeps undelivered outputs; terminal stats delivered FIFO after them; only cancel discards).
- **`test/unit/parakeet-streaming-duplex.test.js`** — regression test that appends chunks and calls `endStreaming()` in one synchronous turn: fails against the old driver (0/3 backlog outputs delivered), passes with the fix.
- Bumped `@qvac/asr-ggml` to `0.3.3` with a changelog entry.

No SDK changes required — the duplex handler and client already behave correctly once events are not dropped.

## Testing

- Unit suite: **168/168 tests, 709/709 asserts**.
- `npm run lint`: clean (prettier + lunte + eslint + typecheck + dts + `check:generated`).
- Regression test verified to fail against the pre-fix driver (via `git stash`) and pass with the fix.
- The C++ change compiles via the CI prebuild workflow (local `bare-make generate` for speech packages needs the post-#312 registry state).

## Follow-up (out of scope)

The streaming path also re-emits overlapping content across chunks (429 words vs the one-shot reference's 301 on the repro audio) due to the rolling-history window — a quality issue, tracked separately per the Asana ticket.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by ogad-tether** [comment]
## End-to-end verification with the ticket's own repro

Ran the QVAC-23758 repro (ported to the asr-ggml API — same three-run design as the script attached to the Asana ticket) against a **local native build of this branch**, using the ticket's `Ai tools.mp3` (93.6 s), Parakeet TDT 0.6B v3 Q8_0, audio fed ~50× realtime with `end()` called immediately after the last chunk. The C++ half was toggled by rebuilding the addon at `HEAD~1` (pre-fix) vs `HEAD` (this fix) — everything else identical.

**Before (this branch's parent, pre-fix):**

```
word counts:   ref=301  strm=70  drain=519
reaches the true end?  strm=false  drain=true
[strm]  …dramatically less important and there are apps now spouting up that can help you talk into your phone or la
```

Immediate `end()` kept only 70 of 301 words — the transcript cuts off ~25 s into a 94 s clip. Only the app-side drain-until-quiet workaround recovered the tail, and that path over-emitted (519 words — the overlap phenomenon the ticket also flagged).

**After (this fix):**

```
word counts:   ref=301  strm=298  drain=298
reaches the true end?  strm=true  drain=true
[strm]  …link in the description there's a way to sign up to our newsletter which gives you great software for less
```

Immediate `end()` now delivers the complete transcript, ending on exactly the reference's final phrase, with no workaround. As a side benefit, in this fast-feed scenario the drained backlog is decoded in large contiguous batches, so the streaming output also stops over-emitting duplicates (298 ≈ ref's 301 rather than 519/429) — the separate overlap issue may still occur under true realtime feeds and remains tracked separately.

Also confirms the C++ change compiles and runs (macOS arm64, Metal).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Review by mexxik — DISMISSED** [review]

File: packages/asr-ggml/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@qvac/asr-ggml",
-  "version": "0.3.2",
+  "version": "0.3.3",
```
Comment (GustavoA1604) [inline packages/asr-ggml/package.json:3]: Don't bump version in the same PR as the fix

File: packages/asr-ggml/CHANGELOG.md [hunk 2]
```diff
@@ -12,6 +12,12 @@ restarts at `0.1.0`; the two pre-merge histories are preserved verbatim as
...
...eaming no longer drops the tail of the transcript when the audio stream ends while the engine still has a buffered backlog (QVAC-23758).** `endStreaming()` joins the native worker only after it has drained all buffered audio, but the drained segments are delivered to JS asynchronously (uv_async); the parakeet driver then cleared the active job and emitted a synthetic `JobEnded` *before* those queued `Output` events arrived, so `_addonOutputCallback` discarded every one of them (`jobId === null`). Apps feeding audio faster than realtime (e.g. the SDK's `transcribeStream` fed from a file, then `end()`) lost most of the un-processed tail. `ParakeetStreamingProcessor` now queues a terminal `RuntimeStats` object (real `audioDurationMs` / `totalSamples`) through the **same FIFO output queue** as the drained segments after `finalize()` — exactly how the whisper engine's `StreamingProcessor` already signals completion — and the JS wrapper waits for that terminal event to flow through the normal output-callback path, so every drained segment is delivered, in order, before the job resolves. The synthetic `JobEnded` remains only as a fallback when no native session existed (`cleaned: false`).
```
Comment (GustavoA1604) [inline packages/asr-ggml/CHANGELOG.md]: Dont add asana ticket

**Review by GustavoA1604 — CHANGES_REQUESTED** [review]
[medium] packages/asr-ggml/src/engines/parakeet/parakeet.ts — line 710 (if (teardown.cleaned) {)

Concurrent endStreaming() calls reintroduce the transcript-tail loss. The first call removes the native session and waits for its queued terminal event; a second call sees cleaned: false, clears _activeJobId, and emits JobEnded before the queued outputs arrive, causing them to be discarded. Cache and return one in-flight end promise, and add a regression test covering duplicate calls.

[low][coding-standards] packages/asr-ggml/addon/src/model-interface/ParakeetStreamingProcessor.cpp — line 256 (// Queue the terminal stats through the SAME FIFO output queue as)

This long block narrates the implementation, contrary to the no-comments standard, and enlarges processLoop(). Extract terminal-stat creation and queuing into a clearly named function, then remove the explanatory block.

File: packages/asr-ggml/package.json (hunk 1 above)
Comment (ogad-tether) [reply packages/asr-ggml/package.json:3]: Reverted — package.json back to 0.3.2 (ac34e1c0f); the changelog entry now sits under [Unreleased] so the release PR can stamp the version.

File: packages/asr-ggml/CHANGELOG.md (hunk 2 above)
Comment (ogad-tether) [reply packages/asr-ggml/CHANGELOG.md]: Removed — dropped the ticket reference from the changelog, and also from the code comments and test names for consistency (ac34e1c0f).

**Comment by ogad-tether** [comment]
Addressed the review in ac34e1c0f:

- **[medium] concurrent endStreaming():** the driver now caches one in-flight teardown promise — a duplicate call joins it instead of hitting the binding again and taking the `cleaned: false` synthetic path. Added a regression test issuing two `endStreaming()` calls in the same synchronous turn while the backlog is undelivered; verified it fails with the guard removed (outputs dropped, duplicate JobEnded) and passes with it.
- **[low] C++ narration block:** extracted into `ParakeetStreamingProcessor::queueTerminalStats()` and removed the explanatory block from `processLoop()`.
- **version bump:** reverted to 0.3.2; changelog entry moved under `[Unreleased]`.
- **Asana reference:** removed from the changelog (and from comments/test names for consistency).

Unit suite 169/169 (714 asserts), lint/typecheck/check:generated clean, native addon recompiled locally.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Review by GustavoA1604 — APPROVED** [review]

**Review by GustavoA1604 — DISMISSED** [review]

**Review by GustavoA1604 — DISMISSED** [review]

**Review by mexxik — APPROVED** [review]

**Review by GustavoA1604 — APPROVED** [review]
